File: src/icicl/curation.py
"""Automatic curation for trajectory databases."""

from icicl.database import TrajectoryDatabase


class CurationManager:
    """Manages automatic curation of trajectory databases.

    Implements exemplar-level curation from the SGICL paper:
    - Tracks which trajectories are retrieved and whether they lead to success
    - Periodically prunes trajectories with low utility scores
    """

    def __init__(
        self,
        database: TrajectoryDatabase,
        threshold: float = 0.3,
        min_retrievals: int = 5,
        curate_every: int = 10,
    ) -> None:
        """Initialize the curation manager.

        Args:
            database: The trajectory database to curate.
            threshold: Utility score threshold below which trajectories are pruned.
            min_retrievals: Minimum number of times a trajectory must be retrieved
                           before it can be considered for pruning.
            curate_every: Run curation after this many successful episodes.
        """
        self._database = database
        self._threshold = threshold
        self._min_retrievals = min_retrievals
        self._curate_every = curate_every
        self._episodes_since_curation = 0

    def maybe_curate(self) -> bool:
        """Check if curation should run and run it if so.

        Returns:
            True if curation was performed, False otherwise.
        """
        self._episodes_since_curation += 1

        if self._episodes_since_curation >= self._curate_every:
            self.curate()
            self._episodes_since_curation = 0
            return True

        return False

    def curate(self) -> list[str]:
        """Run curation to prune low-utility trajectories.

        Returns:
            List of trajectory IDs that were removed.
        """
        removed_ids: list[str] = []

        for trajectory in self._database.get_all():
            metadata = self._database.get_curation_metadata(trajectory.id)

            if metadata is None:
                continue

            if metadata.times_retrieved < self._min_retrievals:
                continue

            if metadata.utility_score < self._threshold:
                if self._database.remove(trajectory.id):
                    removed_ids.append(trajectory.id)

        return removed_ids

    def get_utility_scores(self) -> dict[str, float]:
        """Get utility scores for all trajectories.

        Returns:
            Dictionary mapping trajectory IDs to their utility scores.
        """
        scores = {}
        for trajectory in self._database.get_all():
            metadata = self._database.get_curation_metadata(trajectory.id)
            if metadata:
                scores[trajectory.id] = metadata.utility_score
        return scores

    def get_low_utility_trajectories(self) -> list[str]:
        """Get trajectory IDs that would be pruned if curation ran now.

        Returns:
            List of trajectory IDs with low utility.
        """
        low_utility = []

        for trajectory in self._database.get_all():
            metadata = self._database.get_curation_metadata(trajectory.id)

            if metadata is None:
                continue

            if metadata.times_retrieved < self._min_retrievals:
                continue

            if metadata.utility_score < self._threshold:
                low_utility.append(trajectory.id)

        return low_utility

File: icrl-ts/src/curation.ts
/**
 * Automatic curation for trajectory databases.
 */

import type { TrajectoryDatabase } from "./database";

export interface CurationManagerOptions {
  /** Utility score threshold below which trajectories are pruned (default: 0.3) */
  threshold?: number;
  /** Minimum retrievals before a trajectory can be pruned (default: 5) */
  minRetrievals?: number;
  /** Run curation after this many successful episodes (default: 10) */
  curateEvery?: number;
}

/**
 * Manages automatic curation of trajectory databases.
 *
 * Implements exemplar-level curation from the ICRL paper:
 * - Tracks which trajectories are retrieved and whether they lead to success
 * - Periodically prunes trajectories with low utility scores
 */
export class CurationManager {
  private readonly database: TrajectoryDatabase;
  private readonly threshold: number;
  private readonly minRetrievals: number;
  private readonly curateEvery: number;
  private episodesSinceCuration = 0;

  constructor(database: TrajectoryDatabase, options: CurationManagerOptions = {}) {
    this.database = database;
    this.threshold = options.threshold ?? 0.3;
    this.minRetrievals = options.minRetrievals ?? 5;
    this.curateEvery = options.curateEvery ?? 10;
  }

  /**
   * Check if curation should run and run it if so.
   * @returns True if curation was performed, false otherwise.
   */
  async maybeCurate(): Promise<boolean> {
    this.episodesSinceCuration++;

    if (this.episodesSinceCuration >= this.curateEvery) {
      await this.curate();
      this.episodesSinceCuration = 0;
      return true;
    }

    return false;
  }

  /**
   * Run curation to prune low-utility trajectories.
   * @returns List of trajectory IDs that were removed.
   */
  async curate(): Promise<string[]> {
    const removedIds: string[] = [];

    for (const trajectory of this.database.getAll()) {
      const metadata = this.database.getCurationMetadata(trajectory.id);

      if (!metadata) continue;
      if (metadata.timesRetrieved < this.minRetrievals) continue;

      if (metadata.utilityScore < this.threshold) {
        if (await this.database.remove(trajectory.id)) {
          removedIds.push(trajectory.id);
        }
      }
    }

    return removedIds;
  }

  /**
   * Get utility scores for all trajectories.
   */
  getUtilityScores(): Map<string, number> {
    const scores = new Map<string, number>();

    for (const trajectory of this.database.getAll()) {
      const metadata = this.database.getCurationMetadata(trajectory.id);
      if (metadata) {
        scores.set(trajectory.id, metadata.utilityScore);
      }
    }

    return scores;
  }

  /**
   * Get trajectory IDs that would be pruned if curation ran now.
   */
  getLowUtilityTrajectories(): string[] {
    const lowUtility: string[] = [];

    for (const trajectory of this.database.getAll()) {
      const metadata = this.database.getCurationMetadata(trajectory.id);

      if (!metadata) continue;
      if (metadata.timesRetrieved < this.minRetrievals) continue;

      if (metadata.utilityScore < this.threshold) {
        lowUtility.push(trajectory.id);
      }
    }

    return lowUtility;
  }
}
